src/app/loadout-builder/mod-utils.ts

```diff
@@ -31,3 +41,299 @@ export function someModHasEnergyRequirement(mods: PluggableInventoryItemDefiniti
+  ) {
+    finalEnergy = generalMod.plug.energyCost.energyType;
+  } else if (
+    combatMod?.plug.energyCost &&
+    combatMod?.plug.energyCost?.energyType !== DestinyEnergyType.Any
+  ) {
+    finalEnergy = combatMod.plug.energyCost.energyType;
+  } else if (
+    raidMod?.plug.energyCost &&
+    raidMod?.plug.energyCost?.energyType !== DestinyEnergyType.Any
+  ) {
+    finalEnergy = raidMod.plug.energyCost.energyType;
+  }
+
+  const generalModCost = generalMod?.plug.energyCost?.energyCost || 0;
+  const combatModCost = combatMod?.plug.energyCost?.energyCost || 0;
+  const raidModCost = raidMod?.plug.energyCost?.energyCost || 0;
+
+  const modCost = itemEnergy.used + generalModCost + combatModCost + raidModCost;
+  const energyUsedAndWasted = modCost + itemEnergy.originalCapacity;
+  const energyInvested = Math.max(0, modCost - itemEnergy.originalCapacity);
+
+  return finalEnergy === itemEnergy.originalType ? energyInvested : energyUsedAndWasted;
+}
+
+function getItemEnergyType(
+  defs: D2ManifestDefinitions,
+  item: DimItem,
+  upgradeSpendTier: UpgradeSpendTier,
+  lockItemEnergyType: boolean,
+  bucketSpecificMods?: PluggableInventoryItemDefinition[]
+) {
+  if (!item.energy) {
+    return DestinyEnergyType.Any;
+  }
+
+  const bucketSpecificModType = bucketSpecificMods?.find(
+    (mod) => mod.plug.energyCost && mod.plug.energyCost.energyType !== DestinyEnergyType.Any
+  )?.plug.energyCost?.energyType;
+
+  // if we find bucket specific mods with an energy type we have to use that
+  if (bucketSpecificModType) {
+    return bucketSpecificModType;
+  }
+
+  return canSwapEnergyFromUpgradeSpendTier(defs, upgradeSpendTier, item, lockItemEnergyType)
+    ? DestinyEnergyType.Any
+    : item.energy.energyType;
+}
+
+function energyTypesAreCompatible(first: DestinyEnergyType, second: DestinyEnergyType) {
+  return first === second || first === DestinyEnergyType.Any || second === DestinyEnergyType.Any;
+}
+
+export function getModAssignments(
+  items: DimItem[],
+  mods: PluggableInventoryItemDefinition[],
+  defs: D2ManifestDefinitions | undefined,
+  upgradeSpendTier: UpgradeSpendTier,
+  lockItemEnergyType: boolean
+): Map<string, PluggableInventoryItemDefinition[]> {
+  if (!defs) {
+    return new Map();
+  }
+
+  const bucketSpecificAssignments = new Map<string, PluggableInventoryItemDefinition[]>();
+  const bucketIndependantAssignments = new Map<string, PluggableInventoryItemDefinition[]>();
+  // just an arbitrarily large number
+  let assignmentEnergyCost = 10000;
+
+  for (const item of items) {
+    bucketSpecificAssignments.set(item.id, []);
+    bucketIndependantAssignments.set(item.id, []);
+  }
+
+  const itemSocketMetadata = _.mapValues(
+    _.keyBy(items, (item) => item.id),
+    (item) => getSpecialtySocketMetadatas(item)
+  );
+
+  const generalMods: PluggableInventoryItemDefinition[] = [];
+  const combatMods: PluggableInventoryItemDefinition[] = [];
+  const raidMods: PluggableInventoryItemDefinition[] = [];
```

Comment:
Any ideas of how to make this so we won't have to redo it every time the modslot landscape changes? I'm not sure exactly how, but it seems like we could just try to match up assigments based on the slotting requirements for each mod?

Reply:
Yeah it is possible to make it more generic but I am not sure how to make it so it's completely future proof. For instance raid mods and the nightmare mod will always take up their own mod slot. Anything that goes into the combat socket is interchangeable and general mods always have their own socket.

We could group mods by the socket they will slot into and do an elimination style attempt of finding a home for each mod with a bunch of constraints. We would need to create permutations of each group and then check all combinations of permutations.

The hard thing with making it generic is we need to start creating objects and arrays to track which items have had their socket reserved by a mod as we iterate through them. In process that would blow out our processing times and in the ui I think we would have ui locking issues as chrome will start triggering gc a lot when a user has a full load out of mods.

Finally this is largely a copy of the process mod checking algorithm, I do see some value in keeping them in line from a debugging point of view but as now don't just take the first match there is less value in that.